### ISSUE TITLE
Logstash-input-beats connections still doesn’t close properly

### ISSUE THREAD

**Issue by amirvaza** [logstash-plugins/logstash-input-beats#259]
- Version: 5.0.1 / 3.1.24
- Operating System: ec2 linux
- Config File: follow
- Sample Data:
- Steps to Reproduce: every run

I have around 50 filebeats with this configuration:
<pre>
<code>
output:
  logstash:
    hosts: ["******:10200"]
shipper:
logging:
  level: warning
  to_files: true
  to_syslog: false
  files:
    path: /var/log/filebeat
    name: filebeat.log
    rotateeverybytes: 1000000
    keepfiles: 5
filebeat:
  prospectors:
    - paths:
        - ******
      encoding: utf-8
      input_type: log
      fields:
        type: "***"
        component: "***"
      fields_under_root: true
      multiline:
        pattern: ^[0-9][0-9][0-9][0-9]-[0-9][0-9]
        negate: true
        match: after
</code>
</pre>

And my logstash configuration:
<pre>
<code>
input {
 beats {
    port => 10200
  }
}

output
{
  elasticsearch {
    hosts => ["localhost"]
  }
</code>
</pre>

After a day or two the logstash service stops working since it has too many opened connection and I have to restart the service.
The output of netstat -p shows 16k lines of CLOSE_WAIT
After a search I saw in github a fix related to this issue in version 3.1.24 (and 5.0.1), both versions didn't seem to fix it for me.
It seems that the plugin still doesn't close the connections properly.
Or maybe I am doing something wrong?

My Logstash version is 5.6.2

Thanks.

**Comment by nathanhruby** [comment]
I just hit this today with logstash 2.4.1 and logstash-input-beats 4.0.1 (this is the version I got doing an uninstall and reinstall of the plugin with bin/plugin).

In my case the connections were not closing because one of the output destinations was blocking and causing the pipeline to block (at least, a quick look of the code confirms that this is possible when the events are flushed in the close process)

This is probably something where a error level log event (or circuit breaker) should be emitted if the cnx hash is full of closers for a long time, or if flushes are not completing in a timely fashion.

**Comment by jordansissel** [comment]
the beats input already has a circuit breaker to reject new connections when the internal queue is full. I recommend upgrading Logstash.

**Comment by jordansissel** [comment]
> The output of netstat -p shows 16k lines of CLOSE_WAIT


I'm not sure why this happened. Please provide full details of your deployment (beats config, Logstash config, any source of network interference[1] between beats and logstash).

[1] sources of network interference: firewalls, proxies, load balancers, vips, etc.

**Comment by amirvaza** [comment]
Hi, thanks again for the help, here is the info

// Beats Config
// One windows filebeat:
<pre>
<code>
filebeat:
  prospectors:
    -
      paths:
        - C:\ProgramData\comp\log1.log
      encoding: utf-8
      input_type: log
      fields:
        type: type1
        server: server1
      fields_under_root: true
      multiline:
        pattern: ^\[
        negate: true
        match: after
    -
      paths:
        - C:\ProgramData\comp2\log2.log
      encoding: utf-8
      input_type: log
      fields:
        type: type2
        server: server1
      fields_under_root: true
      multiline:
        pattern: ^\[
        negate: true
        match: after
  registry_file: "C:/ProgramData/filebeat/registry"
output:
  logstash:
    hosts: ["mylog.mydomain.com:10200"]
shipper:
logging:
  level: warning
  to_files: true
  to_syslog: false
  files:
    path: "C:/ProgramData/filebeat"
    name: filebeat.log
    rotateeverybytes: 1000000
    keepfiles: 5
</code>
</pre>


100s of linux filebeats (Changing by autoscale, pretty much look the same):
<pre>
<code>
output:
  logstash:
    hosts: ["mylog.mydomain.com:10200"]
shipper:
logging:
  level: warning
  to_files: true
  to_syslog: false
  files:
    path: /var/log/filebeat
    name: filebeat.log
    rotateeverybytes: 1000000
    keepfiles: 5
filebeat:
  prospectors:
    - paths:
        - /var/log/comp/log1.log
      encoding: utf-8
      input_type: log
      fields:
        type: "type1"
        component: "comp1"
        server: "serverX"
      fields_under_root: true
      multiline:
        pattern: ^[0-9][0-9][0-9][0-9]-[0-9][0-9]
        negate: true
        match: after

    -
       paths:
         - /var/log/comp2/log1.log
       encoding: utf-8
       input_type: log
       fields:
         type: "type2"
         server: "serverX"
       fields_under_root: true
       multiline:
         pattern: ^[0-9][0-9][0-9][0-9]-[0-9][0-9]
         negate: true
         match: after

</code>
</pre>


Logstash Config
<pre>
<code>
input {
 beats {
    port => 10200
  }
}
 
filter {

  if [type] in ["type1", "type2", "type3"] {

    # Parse the message
    grok {
      overwrite => "message"
      match => { "message" => "\[%{TIMESTAMP_ISO8601:timestamp}\] (?<thread>.*)@%{NUMBER:pid} %{WORD:user} %{WORD:level} *%{GREEDYDATA:message}" }
    }

    # Repalce the date with the timestamp in the message
    date {
      match => ["timestamp", "ISO8601"]
      target => "@timestamp"
    }

    # Add field system, this is the type we got from the server
    # Remove fields
    mutate{
      add_field => {"system" => "%{type}"}
      remove_field => ["tags", "source", "offset", "input_type", "count", "beat", "@version", "timestamp", "type"]
    }
  }

  else if [type] in ["type4", "type5"] {
  	..... 
  }
  .
  .
  .


  // Windows
  else if [type] in ["type9", "type10"] {

    # Parse the message
    grok {
      overwrite => "message"
      match => [ "message", "\[%{TIMESTAMP_ISO8601:timestamp}\] *(?<thread>[^ ]*) *%{WORD:level} *%{USERNAME:class} *(?<user>[^ ]*) *\r\n *Exception: (?<exception>(.*\r\n)*) Message: %{GREEDYDATA:message}"]
    }

    # Repalce the date with the timestamp in the message
    date {
      match => ["timestamp", "ISO8601"]
      target => "@timestamp"
    }

    # Add field system, this is the type we got from the server
    # Remove fields
    mutate{
      add_field => {"system" => "%{type}"}
      remove_field => ["tags", "source", "offset", "input_type", "type", "count", "beat", "@version", "timestamp"]
    }

    # changing "WARN" to "WARNING"
    ruby {
         code => "event.set('level', {'WARN' => 'WARNING'}[event.get('level')] || event.get('level'))"
    }
  }

}

output
{
  elasticsearch {
    hosts => ["localhost"]
  }
}
</code>
</pre>

**Comment by soulis-siluos** [comment]
Has there been any progress on this issue?

**Comment by amirvaza** [comment]
Unfortunately not from my side.

@jordansissel

**Comment by heanssgen-troy** [comment]
We have also experienced this in production.

It seems to only affect the beats input plugin, as any logstash not receiving beats data is unaffected. Has there been any headway in figuring out why we are leaking sockets?

**Comment by jsoref** [comment]
So, in my environment, each system connects via an autossh managed ssh tunnel
filebeat 5.5.2 (running on ubuntu 16.04.3)
connects to->localhost:5044
ssh listens on :5044 and forwards to -> logstash:5044
logstash has logstash-input-beats-3.1.24-java
```sh
root@logstash-live:~# for a in CLOSE_WAIT FIN_WAIT2; do grep -c $a /tmp/lsof-5044; done
378027
1064
root@logstash-live:~# lsof -n -P |grep 5044 > /tmp/lsof-5044
root@logstash-live:~# grep ^sshd /tmp/lsof-5044 |sed -e 's/.*\(TCP.*\):.*-/\1:x/'| uniq -c
   1064 TCP [::1]:x>[::1]:5044 (FIN_WAIT2)
root@logstash-live:~# grep -v ^sshd /tmp/lsof-5044 |perl -ne 'next unless s/^(\S+).*(TCP.*?):5044.*/$1 $2:x/;print'| sort|uniq -c|sort -n
      1 [main]>wo TCP *:x
      5 Api\x20We TCP *:x
      5 [main]-pi TCP *:x
      7 LogStash: TCP *:x
      8 [main]<be TCP *:x
     13 java TCP *:x
   9693 [main]>wo TCP [::1]:x
  48465 Api\x20We TCP [::1]:x
  48465 [main]-pi TCP [::1]:x
  67851 LogStash: TCP [::1]:x
  77544 [main]<be TCP [::1]:x
 126009 java TCP [::1]:x
```

ssh eventually reports:
```
**Unmatched Entries**
error: socket: Too many open files : 1 time(s)
error: connect to 127.0.0.1 port 41514 failed: Too many open files : 1 time(s)
error: connect_to localhost: unknown host (Too many open files) : 1 time(s)
```
perhaps:
https://stackoverflow.com/questions/28875406/why-am-i-seeing-lots-of-sockets-in-close-wait-status-when-webservice-stops-worki

**Comment by jordansissel** [comment]
Do you experience this problem with direct connection to logstash (no
proxy/tunnel/etc)?

**Comment by original-brownbear** [comment]
@jordansissel just gave the code a quick read here, it seems we're never closing channels on exceptions in the main loop (I could be way off in the below though, sorry for the noise in that case :))?

I can see https://github.com/logstash-plugins/logstash-input-beats/blob/master/src/main/java/org/logstash/beats/Server.java#L137

which just does:

```java
        @Override
        public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
            logger.warn("Channel initializer");
            this.message.onChannelInitializeException(ctx, cause);
        }
```

and only seems to call https://github.com/logstash-plugins/logstash-input-beats/blob/master/src/main/java/org/logstash/beats/MessageListener.java#L65:

```
    /**
     * Called when a error occur in the channel initialize, usually ssl handshake error.
     *
     * @param ctx
     * @param cause
     */
    public void onChannelInitializeException(ChannelHandlerContext ctx, Throwable cause) {
        logger.debug("onException");
    }
```

maybe we're eventually leaking a bunch of fds on Channels that errored out and never got closed?

**Comment by heanssgen-troy** [comment]
Forget not closing channels on exceptions, we are not closing channels at all.

If we add the following code to the beats plugin:
`
def onNewConnection(ctx)
      input.logger.error("ARRRR, we be gettin' a connection maties.")
      register_connection(ctx)
      increment_connection_count()
end

def onConnectionClose(ctx)
      input.logger.error("ARRRRR, we be closin' a connection maties.")
      unregister_connection(ctx)
      decrement_connection_count()
end
`
We can see that when the pipeline is completely stalled (unable to process ANY events), the connection close event is never fired. It appears that we might be blocking on message processing by way of https://github.com/logstash-plugins/logstash-input-beats/blob/master/lib/logstash/inputs/beats/codec_callback_listener.rb:

`
def process_event(event)
      @transformer.transform(event, @hash)
      @queue << event
end
`

**Comment by original-brownbear** [comment]
@heanssgen-troy yea looks like we're getting warmer there, thanks!

Will try to fix this now.

**Comment by original-brownbear** [comment]
@heanssgen-troy 

>  It appears that we might be blocking on message processing by way of https://github.com/logstash-plugins/logstash-input-beats/blob/master/lib/logstash/inputs/beats/codec_callback_listener.rb:

jup you're right on this point :) Just tracked this down, the problem is that our `@codec`s aren't thread safe  => can't just share them across all Netty threads like we do atm.

This must be fixed the same way I fixed it in Kafka in https://github.com/logstash-plugins/logstash-input-kafka/pull/211.

**Comment by heanssgen-troy** [comment]
There is also another offending piece of code with the same synchronization errors as the codec. We can see this code in https://github.com/logstash-plugins/logstash-input-beats/blob/master/lib/logstash/inputs/beats/message_listener.rb

```
event = LogStash::Event.new(hash)
@nocodec_transformer.transform(event)
@queue << event
```

Both calls will block when the pipeline is stalled, since @queue << event is just a simple alias for a java backed Queue.push().

**Comment by original-brownbear** [comment]
@heanssgen-troy true, but the pipeline should in fact never stall for a long period of time. Also, there isn't really an alternative to blocking on a full pipeline is there? (apart from dropping events) It's just the unfortunate physics of things that eventually back-pressure will have to be applied I guess?

**Comment by heanssgen-troy** [comment]
@original-brownbear In reality though, outages DO happen and on a regular basis. We should not block the entire processing pipeline because a few beats inputs are stalling.

Take our configuration for example. We have a series of data providers that ship their information straight to Elasticsearch for later searching. We also have a configuration whereby we use Logstash as an intermediate shipper in order to process the data and forward it to RabbitMQ (and from there another application).

The beats data is low priority in this case, but the act of the Elasticsearch output stalling will back the entire pipeline up preventing our much more powerful custom application from doing realtime processing.

In this case, we really don't care if the beats data is processed (we can afford to drop that), but the UDP data is of utmost priority (Before you say TCP, this data is coming in VIA multicast). Having the ability to configure the pipeline to drop events would be very useful.

Below is a visualization of what im talking about:

```
Beats Input -> Logstash -> Elasticsearch
UDP Input -> Logstash -> RabbitMQ -> Custom Application
```

**Comment by original-brownbear** [comment]
@heanssgen-troy got it! As a matter of fact this feature (we're talking about it in the context of just putting a timeout on pushes to the main Queue) is being discussed in LS core already :) So this is incoming as a feature eventually.

**Comment by heanssgen-troy** [comment]
@original-brownbear Thanks for the heads up :)

It looks like this issue is already being addressed in https://github.com/logstash-plugins/logstash-input-beats/pull/272. We had first contacted the Elastic Team regarding this issue, and it appears that headway is being made.

Thanks for everyone's help!

**Comment by ph** [comment]
@heanssgen-troy We plan to have a timeout or allow the queue to return a future which will allow us to better handle backpressure. Is currently OK to block, the problem is when this situation happens FB need to be notified to keep the connection open but wait before sending any data. There was a bug in the logic which prevented that, make FB client raise a "read io timeout" and reconnecting effectively increasing the connection count.

**Comment by ph** [comment]
One scenario we could do with better semantic on the queue is the following:

FB -> LS (Hang) -> retry X time -> Fails -> FB pick up another LS host.

**Comment by ph** [comment]
```
Beats Input -> Logstash -> Elasticsearch
UDP Input -> Logstash -> RabbitMQ -> Custom Application
```

is Logstash the same instance?

**Comment by ph** [comment]
@heanssgen-troy like @armin said the conn was not correct closed on exception, this was a bit of  a rabbit hole.

**Comment by ph** [comment]
> @original-brownbear In reality though, outages DO happen and on a regular basis. We should not block the entire processing pipeline because a few beats inputs are stalling.

The semantic of blocking applies to all the inputs not just the beats we currently consider the pipeline as a whole unit, but in LS 6 there is way around this, you can run mutliple pipeline in parellel with different requirements and settings. So this would help in the case that one execution can block and the other should not be affected by this outage.

**Comment by amirvaza** [comment]
Hey all,

Any news?

Thanks!

**Comment by dwjvaughan** [comment]
+1 on this, having the same issue with beats 5.6.5 and logstash 5.6.4.

**Comment by original-brownbear** [comment]
@ph assigning you here as discussed a while ago :)

**Comment by amirvaza** [comment]
Hey guys, sorry to interrupt, any progress on this?

**Comment by original-brownbear** [comment]
@ph (more likely: @robbavey since you're on the memory improvements) One of you guys should incorporate https://github.com/logstash-plugins/logstash-input-beats/pull/274 into their work imo. Last time I checked the fixes in there are still missing from `master`.

**Comment by eberkut** [comment]
Just confirming that I'm still seeing this issue almost daily with filebeat 5.6.7, logstash 5.6.7 and logstash-input-beats 3.1.26.

**Comment by jakelandis** [comment]
I believe this is resolved with

Logstash 6.x - `5.0.9`
Logstash 5.x - `3.1.28`

@robbavey - can you confirm ?

**Comment by tomsommer** [comment]
Seems to be fixed in 8.0.10 -- Jesus guys, this bug is nasty - Logstash should roll a new version/bundle ASAP.

**Comment by MorrieAtElastic** [comment]
Can I get confirmation that this issue is fixed in logstash-inputs-beats 3.1.28? I'm very hesitant to ask my customer to make yet another configuration change as the previous 2 upgrades she made did not solve the problem.

**Comment by MorrieAtElastic** [comment]
Can I get confirmation that this problem is fixed in logstash-input-beats-3.1.28? There seems to be some question as to whether this bug has been resolved.

**Comment by robbavey** [comment]
This bug should be resolved as of beats inputs versions `5.0.11` and `3.1.30`

**Comment by MorrieAtElastic** [comment]
Customer on whose behalf  https://github.com/elastic/support-dev-help/issues/2958 was filed now reports no interruptions of processing after running multiple logstash processes continuously for 15 days on logstash-input-beats  3.1.30. 

So I comfortable that this problem is finally resolved in 3.1.30.
